lib/s3/storage.go

```diff
@@ -38,20 +39,39 @@ type Storage struct {
 	s3       *s3.S3
 }
 
-// List reads the path content
-// Note: it is not better performant than Walk because it still
-// itterates over all objects in a path
-func (s *Storage) List(path string) ([]string, error) {
+// List reads the path content or prefixes.
+func (s *Storage) List(path string, options ...map[string]interface{}) ([]string, error) {
 	var result []string
-	res, err := s.s3.ListObjects(&s3.ListObjectsInput{
+
+	input := s3.ListObjectsInput{
 		Bucket: aws.String(s.bucket),
 		Prefix: aws.String(path),
-	})
+	}
 
-	if err != nil {
-		return result, err
+	if len(options) > 0 {
+		if options[0]["delimiter"] != nil {
+			input.SetDelimiter(options[0]["delimiter"].(string))
+		}
 	}
 
+	err := s.s3.ListObjectsPages(
+		&input,
+		// handle bulks of 1000 keys
+		func(res *s3.ListObjectsOutput, _ bool) bool {
+			if input.Delimiter != nil {
+				result = append(result, prefixes(res)...)
+			} else {
+				result = contents(result, res)
+			}
+
+			return true
+		},
+	)
+
+	return result, err
+}
+
+func contents(result []string, res *s3.ListObjectsOutput) []string {
```

Comment:
Probably should make this function more `pure` as well to make sure it's consistent. I mean:
```go
func contents(res *s3.ListObjectsOutput) []string
```
Then in list function:
```go
result = append(result, contents(res)...)
```

Reply:
done.